### DISCUSSION TITLE
GitHub Actions "release" trigger should trigger on draft releases

### DISCUSSION THREAD

**Discussion by dwightguth** [community/community#7118] [Actions]
If you look at https://docs.github.com/en/actions/learn-github-actions/events-that-trigger-workflows#release, you will see that it says "The release event is not triggered for draft releases". This is really weird, however, because it has both the "published" and "created" activity types. "created" triggers on any new release, and "published" triggers on any new release that is not a draft. There is no reason to have both of these types if you are silently discarding all draft releases. If I wanted to create a workflow that ran whenever a release was published, I would subscribe to the "published" activity. The "created" event fires when a draft release is created, and yet subscribing to this event doesn't trigger anything in this case? The behavior of the workflow trigger is inconsistent with the releases API for no particular reason.

Yes, of course I can use a pre-release instead, but pre-releases are publicly visible, which is undesirable when what you really want is a draft. It would be nice if this functionality just worked the way you would expect rather than having this weird gotcha.

**Comment by Marcin-Lach** [comment]
Would be nice if triggering workflows after a draft release was created.
In my case I have a white-label project and multiple customer-specific projects (each in separate repository). I wanted to have a process that:
1. creates release of white-label project as draft after pushing tags
2. dispatch workflows in each customer-specific project that will run tests on draft release
3. if all tests succeed, publish white-label project

**Comment by Endermanbugzjfc** [comment]
Neither this is fixed nor the type description is changed after a year.

**Comment by cliffplaysdrums** [comment]
Same Our use case is that we have a subset of tests that are expensive and only want to run them for new releases, but we want the tests to pass before we publish the release. Triggering these tests on the creation of a draft release seems like the best way to go about it.

**Comment by magurotuna** [comment]
+1 for this. We have a test script that checks if the version specifier hard-coded in a file matches the version being released. It would be really nice if we could run this test when a draft release is created.

**Comment by movermeyer** [comment]
In our case, we release so infrequently that we always need to perform tests of our release pipeline to make sure that things are still working (Something is always broken due to bit rot. You know how software is).

My goal was to have the user create a Draft Release, which would trigger the pipeline to run and upload to the [test PyPI server](https://test.pypi.org/).

Once the release pipeline has been verified as working, the user would publish the release and the same pipeline would upload to the production PyPI server instead.

However, since GitHub `release: create` events aren't created for Draft Releases, this process doesn't work.

-----

As a cludgy workaround, I have tried to hack together a `workflow_dispatch` event with `inputs` that mimic those of the `release` event. This creates two UX flows (one for release pipeline tests and one for releases) and makes the workflow itself harder to maintain, since it has to try to harmonize the different trigger mechanisms.

**Comment by lbruun** [comment]
>  It would be nice if this functionality just worked the way you would expect rather than having this weird gotcha.

It would be nice if it just worked as described in the [WebHook events doc](https://docs.github.com/en/webhooks/webhook-events-and-payloads#release). I haven't tried with a Webhook, but as described in the link, it seems that the creation of a Draft Release from the UI would trigger a webhook event, but not a pipeline event. Either that or the docs are incorrect.

**Reply by justinmchase** [reply]
The webhook / app does indeed get the event, its only github workflows that don't. In fact that is the best workaround, build an app which listens to the event and does all the work in the app.

**Comment by couling** [comment]
I don't think the need is to have draft trigger a release, but to get all the build and testing done **before** publishing.

One workaround is to invert the order.  Have a manually triggered action [create the release](https://github.com/marketplace/actions/create-release) (possibly in draft).  If the build and tests fail, the release never exists.  If it succeeds, it is automatically documented by creating a release.

**Reply by justinmchase** [reply]
Its specifically to set build versions in code to match the version of the draft release.

**Comment by dietergeerts** [comment]
I agree that the use-case of testing is not a good one, as you can build your workflow in such a way that testing is done before creating the release. We do this by running the tests on the branch that gets automatically created by the release-please action. That said, I also believe that having a trigger on a draft release created is needed for another scenario:

- the draft release is created (by release-please in our case)
- the release.yml workflow triggers and will do:
  - create the assets and upload it to the release
  - publish the release as all assets are now available

This will make sure that when other tools pick up the release, it already has all the assets, because otherwise we could have the case that there is a download of assets that aren't there yet. Also, the asset creation could fail and we'll have to look into that first.

**Reply by couling** [reply]
I've certainly seen senarious where all of the testing passed and the moment the build pipeline ran to generate the release it fell flat. We then had to pull the release and do a lot of cleanup.

**Reply by JonasCir** [reply]
I have exactly the same use case! I only want to fully release once all artifacts are attached.

**Reply by justinmchase** [reply]
For me the use case is not to test but to update the versions in the code to match the release before its officially released.

**Comment by mosteo** [comment]
My personal workaround for this one is to test-release in a personal fork of the official repo. If it succeeds it also means I have a backup of all artifacts that I could upload manually to the official release if that one were to fail due to some flaky test.

Still, it would be much better just to be able to change from draft to proper and not have to re-run all the release pipeline, with the added suspense.

**Comment by Sammcb** [comment]
I also agree that creating a draft release to trigger asset creation/attachment is a valuable use case.  I think this would be especially useful with the new [immutable releases](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/immutable-releases) feature entering public preview. It would be great if I could create a draft release, have the artifacts generated and attached via a triggered action, and then publish the immutable release once I verify everything looks correct.

**Reply by emmercm** [reply]
This is the exact process I was hoping to use with my project.

**Reply by samuelss1996** [reply]
Same here!

**Reply by samuelss1996** [reply]
In case anyone is looking for something similar to this workflow, I have a workaround: create tags and let the draft releases be created automatically.

The idea is to create the tag at the specific commit you want the release using:

```
git tag v.0.1.0
git push origin v.0.1.0
```

And then have a worflow that is triggered on tag creation that builds the artifacts and automatically creates the draft release with the attached artifacts. This problem is solved by using:

https://github.com/softprops/action-gh-release

For me it worked like a charm. Hope it helps anyone.

**Comment by thomasturrell** [comment]
I’d really value support for triggering Actions workflows when a draft release is created.

Right now, the release event only fires when a release is published, not when a draft is created. For many projects this is a real limitation, because the most appropriate workflow for a product owner or release manager is to use the GitHub Release UI as the primary interface for preparing a release. Creating and editing a draft release is the natural place where the human decides the version number, curates the changelog, and prepares the notes, and it often needs to happen before any build or verification workflow runs.

This becomes even more important with the introduction of immutable releases, where the first published version of a release must be correct and complete. In that model, I want the process to look like this:

1. A human creates a draft in the release UI
2. That action triggers a workflow to:
    - validate the draft release (e.g. version number)
    - set the version number in the code (e.g. `pom.xml`, `package.json`)
    - create a commit for the version change
    - build and test the version
    - produce and upload assets to the release
    - create a tag
3. Only after the workflow completes successfully should publishing become possible 

This mirrors how code integrity is protected today: pull requests, branch protection, required reviews, and required checks. Release integrity can be just as important as code integrity. For many projects, the release artefacts, versioning, and notes matter just as much as the code they originate from.

Having a "draft created" trigger would allow projects to enforce a proper release pipeline and verify everything before a release becomes public. It would also allow teams to mandate the use of draft releases, making the GitHub Release UI the natural, supported entry point for release creation, rather than relying on custom scripts, manual `workflow_dispatch` triggers, or complex workarounds where a workflow fires on `release.published`, copies the release, deletes it, and re-creates it just to run the required automation.

**Reply by justinmchase** [reply]
This is absolutely correct and this is how we use it today. Of course you have to implement an app to handle this event but it works exactly as you described it. Adding this event to actions would be way simpler since you could implement it all with actions instead of a custom app.

**Comment by woodruffw** [comment]
I was just bitten by this. Like others, I expected this to work given that GitHub has shipped immutable releases, which _require_ managing release uploads while still in the 'draft' state.

I think it's pretty frustrating that the only alternative is a tag push event, which means having the workflow manage the entire release state rather than just the uploads it's actually responsible for.

**Comment by matmair** [comment]
Especially when looking at this https://docs.github.com/en/code-security/concepts/supply-chain-security/immutable-releases#best-practices-for-publishing-immutable-releases recommendation to use drafts it would be great to actually have a trigger that does this; workflow_dispatch just feels like a little hacky

**Comment by lima-limon-inc** [comment]
Hello, have there been any updates on this feature? If backwards compatibility is an issue, maybe a new `drafted` keyword/event could be added instead that is specifically design to trigger on `draft` creation.
